[Woo POS] [Cash & Receipts] UI updates from testing

iamgabrielma opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #14910

## Description
This PR addresses some of the UI updates brought up from testing ( pdfdoF-6dp-p2 ) in order to bring the iOS UI/UX closer to what Android already has implemented. I'm intentionally leaving out some of the items brought up in testing, as we still have no clear consistent designs, and some of the items raised are platform-specific preferences. 

If any, these can be handled outside of the project as needed.

## Changes

- Updates the "change due" text colour from "success green" to "gray":

<img width="208" alt="Screenshot 2025-01-20 at 17 17 31" src="https://github.com/user-attachments/assets/8e36294b-6a63-4897-974c-e94805f47a84" />

- Adjust the "mark payment as complete" button height dynamically when keyboard is presented/hidden:

https://github.com/user-attachments/assets/78bb1e16-1313-46cf-b7ef-50025d60c82d

- Update the green colour used on success checkmark when dark mode, and moved it to POS constants file.

| Light | Dark |
|--------|--------|
| ![Simulator Screenshot - iPad Air 11 - iOS 17 5 M2 - 2025-01-21 at 11 31 06](https://github.com/user-attachments/assets/ce8afe7e-f5f5-402a-9b32-98f3bc0dc1a5)  |  ![Simulator Screenshot - iPad Air 11 - iOS 17 5 M2 - 2025-01-21 at 11 41 00](https://github.com/user-attachments/assets/dc7c72ee-481c-4f86-8fdf-1255c2bbaafa) | 

## Testing information
* Enable `acceptCashForPointOfSale` flag
* Go through a POS transaction, select payment with cash.
* Observe that:
  * When adding a higher amount than the order total, the "change due" text colour is no longer green
  * Showing/hiding the keyboard makes the "mark as complete" button move along, rather than being partially hidden or being static
  * Upon successful payment, in dark mode, the success checkmark circle is dark green.

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

iamgabrielma committed:
update change due text color

iamgabrielma committed:
button dynamic spacing when button shown

iamgabrielma committed:
Extract KeyboardObserver

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/235">21.5</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

iamgabrielma committed:
add preview

iamgabrielma committed:
Add keyboard observer to receipts view

iamgabrielma committed:
Move circle bg color to POS constants

success text color happened to be different, but we no longer use it. Let’s move the circle success color to POS constants so we do not have different tones all around.

We still keep the same values for light-dark modes

iamgabrielma committed:
update dark mode color for circle success

iamgabrielma committed:
Merge branch 'trunk' into task/14910-pos-cash-testing-feedback-ui-updates

iamgabrielma committed:
clean up keyboard observer

iamgabrielma commented on WooCommerce/Classes/POS/ViewHelpers/KeyboardObserver.swift:
```diff
@@ -0,0 +1,35 @@
+import SwiftUI
+import Combine
+
+final class KeyboardObserver: ObservableObject {
```
Happy to get some feedback on this `KeyboardObserver`. I wasn't quite sure how to handle the dynamic height for the buttons when keyboard is on/off or when we present messages like error/change due. Right now it seems to do the job but I'm unaware if there are better solutions 🤔

joshheald replied:
I don't know... but you might not need any of it. I've certainly never had to worry about it in SwiftUI before.

[This article](https://www.fivestars.blog/articles/swiftui-keyboard/) suggests that we should just be able to rely on SwiftUI to handle it for us. We can wrap it in a scroll view and it might just work...

At larger dynamic type sizes we'll need one anyway; the button gets hidden by the keyboard at the moment.

joshheald committed:
Merge branch 'trunk' into task/14910-pos-cash-testing-feedback-ui-updates

joshheald commented on WooCommerce/Classes/POS/Presentation/Reusable Views/POSSendReceiptView.swift:
```diff
@@ -15,6 +16,14 @@ struct POSSendReceiptView: View {
         EmailFormatValidator.validate(string: textFieldInput)
     }
 
+    private var keyboardHeight: CGFloat {
+        if keyboardObserver.keyboardHeight < Constants.keyboardShownButtonSpacing {
+            return Constants.keyboardShownButtonSpacing
+        } else {
+            return Constants.keyboardHiddenButtonSpacing
+        }
+    }
```
This doesn't return the height of the keyboard, so it's a confusing name...

joshheald commented on WooCommerce/Classes/POS/Presentation/Reusable Views/POSSendReceiptView.swift:
```diff
@@ -15,6 +16,14 @@ struct POSSendReceiptView: View {
     }
 
+    private var keyboardHeight: CGFloat {
+        if keyboardObserver.keyboardHeight < Constants.keyboardShownButtonSpacing {
```
I don't really understand the intent of this comparison... could you explain a little?

Why would the keyboard being taller than the shown button spacing mean that we should return the shown button spacing? Surely there's some other threshold to decide that?

At the moment, there are some unexpected changes of this spacing at times, and it might be because of this comparison...

joshheald commented on WooCommerce/Classes/POS/Presentation/Reusable Views/POSSendReceiptView.swift:
```diff
@@ -56,8 +65,11 @@ struct POSSendReceiptView: View {
+                    .padding(.bottom, Constants.errorMessagePadding)
             }
 
+            Spacer().frame(height: keyboardHeight)
```
Perhaps you could make a `viewModifier` which you use on this spacer, so we don't declare all this keyboard observation twice?
```suggestion
            Spacer()
                .keyboardResponsiveHeight(
                    withLargeKeyboard: Constants.keyboardShownButtonSpacing, 
                    withoutKeyboard: Constants.keyboardHiddenButtonSpacing)
```

joshheald commented on WooCommerce/Classes/POS/Presentation/Reusable Views/POSSendReceiptView.swift:
```diff
@@ -86,6 +98,7 @@ struct POSSendReceiptView: View {
         }
         .padding([.horizontal, .bottom])
         .animation(.easeInOut, value: errorMessage)
+        .animation(.easeInOut, value: keyboardObserver.keyboardHeight)
```
I'm not sure how nicely this animation will play if you do the separate viewmodifier though. It may still be possible, but check before you go too deep.

joshheald reviewed (commented):
Some suggestions to try inline. For summary:

1. Try removing all the observation and just wrapping it in a ScrollView.
2. If you still need the observation, try encapsulating it in a view modifier rather than duplicating across the two views.
3. If that doesn't work, just tidy up the keyboard height naming and logic a bit...

I found that it's broken in various ways with larger dynamic text sizes:

Even at 90%, the disconnected card reader view has truncated text now
![IMG_0299](https://github.com/user-attachments/assets/570d35b8-b6df-4dae-b0e3-6bf3ee50b067)

When you increase the text size, the cash button gets pushed off screen – here's 135%
![IMG_0300](https://github.com/user-attachments/assets/dbac016d-cfbd-4172-a78f-c154460c2d68)

When you get up to the very biggest sizes, it's not accessible or visible at all:
![IMG_0302](https://github.com/user-attachments/assets/e8fcd222-5254-4478-9b7b-d4f2db83be82)

On the cash and email entry screen, it gets pushed behind the keyboard, if the keyboard is full width not floating. You can still usually get to it by pressing enter, but it would be nicer in a scrollview, I think.

iamgabrielma commented:
Thanks for testing Josh! 

> I found that it's broken in various ways with larger dynamic text sizes

While I address the keyboard observer feedback and other items, can you check if the screenshots/testing were made before or after merging trunk? Initially I thought differences you see could be from iOS17 vs 18 discrepancies, but I'm seeing the views behaving "looking fine" both using iOS17.6.1 (physical) and  iOS18 (simulator) and very different from your screenshots.

Eg iOS 17.6.1 recording on physical device with 135% max font size:

https://github.com/user-attachments/assets/6048cebf-606b-4f4b-9b6f-4c99c5906ef8

And this is with massive sizes also on iOS17.6.1 in physical device.

https://github.com/user-attachments/assets/e9a8b780-9639-4d2e-b688-65ba48f20b89

joshheald committed:
Merge branch 'trunk' into task/14910-pos-cash-testing-feedback-ui-updates

joshheald commented:
> While I address the keyboard observer feedback and other items, can you check if the screenshots/testing were made before or after merging trunk? Initially I thought differences you see could be from iOS17 vs 18 discrepancies, but I'm seeing the views behaving "looking fine" both using iOS17.6.1 (physical) and iOS18 (simulator) and very different from your screenshots.

Ah, I didn't realise this was fixed on trunk. I've merged trunk to this PR, and it doesn't have nearly so many issues with dynamic type after that.

The only remaining one I see is that the CTA button for `Mark payment as complete` gets smaller in the largest accessibility text sizes, which isn't great.

iamgabrielma committed:
Wrap cash view in ScrollView. Remove keyboard observer

iamgabrielma committed:
Merge branch 'task/14910-pos-cash-testing-feedback-ui-updates' of https://github.com/woocommerce/woocommerce-ios into task/14910-pos-cash-testing-feedback-ui-updates

iamgabrielma committed:
remove keyboard observer from receipts view and adjust conditional padding

iamgabrielma committed:
Use AdaptiveButtonPaddingModifier for button padding

iamgabrielma committed:
linting

iamgabrielma commented:
Thanks for double-checking! I'm glad wasn't something off with iOS versions 😅 

Since the changes I've also seen that we don't need the keyboard observer, as you mentioned in the comments, and a simple ScrollView for the cash view is enough, just to be sure that all the vertical components will be accessible if they happen to be partially hidden due to sizing. I've updated this here: https://github.com/woocommerce/woocommerce-ios/pull/14919/commits/5f9b8b9fa07b1de9a9a26c9dd82a9a0eaee890d0 and removed entirely the keyboard observer on https://github.com/woocommerce/woocommerce-ios/pull/14919/commits/bc427fb18c57a47fd39759c1ab1869b11808c5a6

> The only remaining one I see is that the CTA button for Mark payment as complete gets smaller in the largest accessibility text sizes, which isn't great.

Right, applying `conditionalPadding` works fine when we want to remove padding entirely based on font size, but doesn't work that well for smooth size transitions, specially between bigger-but-non-accessible and accessible sizes. I've followed your idea of making a view modifier here https://github.com/woocommerce/woocommerce-ios/pull/14919/commits/3c8757ccccf44231d5000ab2484b66dcfa62506b and adapt the size of the button based on `dynamicTypeSize `, which is also limited to ~60% of its default size as we set its upper bound via `.dynamicTypeSize(...DynamicTypeSize.accessibility1)` so it cannot go smaller. 

Here's for both cash and receipts views high the highest font size and error messaging:

| Email receipts | Cash payments |
|--------|--------|
| ![Simulator Screenshot - iPad Air 11 - iOS 17 5 M2 - 2025-01-23 at 14 41 46](https://github.com/user-attachments/assets/c6cda344-bfeb-4777-b29b-9303d6e007ac) | ![Simulator Screenshot - iPad Air 11 - iOS 17 5 M2 - 2025-01-23 at 15 11 13](https://github.com/user-attachments/assets/7005c98d-a9d1-4c88-90c9-e5c1c9a9d38a) |

joshheald commented on WooCommerce/Classes/POS/Presentation/PointOfSaleCollectCashView.swift:
```diff
@@ -25,81 +25,84 @@ struct PointOfSaleCollectCashView: View {
+
+                FormattableAmountTextField(viewModel: textFieldViewModel, style: .pos)
+                    .focused($isTextFieldFocused)
+                    .dynamicTypeSize(...DynamicTypeSize.accessibility1)
```
Is limiting the size the only option here?

I'd expect people to only select the giant sizes if they really need them; limiting the range available means we're not fully supporting those users.

iamgabrielma replied:
With the keyboard on the way, is the best I found to make it work both for iPads and iPad mini in large accessibility sizes. I think there are alternatives like we do with the header (composing it horizontally rather than vertically) and we could do the same for textfield + button and put them side by side on larger sizes rather than force them to stack vertically. This could be done with [ViewThatFits](https://developer.apple.com/documentation/swiftui/viewthatfits) for example 🤔 

I think we haven't (or at least personally I have not) thought about those screens with accessibility in mind when were designed, and we could have a different version that adapts nicely with larger sizes. I've raised this with design as something we can do going forward, and I'll definitely add it to my list as well when we explore features and UI.

I like the solution you propose for iOS17+, I've spent some time today testing it and seems to work nicely. I've decided to [log it separately](https://github.com/woocommerce/woocommerce-ios/issues/14970) thought and not make the changes in this PR, as we need to add several conditional modifiers to make it work with iOS16 and iOS17+ and I'm not confident of or having the time to test this properly before the code freeze in a couple of hours.

joshheald commented on WooCommerce/Classes/POS/Utils/AdaptiveButtonPaddingModifier.swift:
```diff
@@ -0,0 +1,33 @@
+
+    private func adjustablePadding(for size: DynamicTypeSize, defaultButtonPadding: CGFloat) -> CGFloat {
+        switch size {
+        case .xxxLarge:       return defaultButtonPadding * 0.7
+        case .xxLarge:        return defaultButtonPadding * 0.8
+        case .xLarge:         return defaultButtonPadding * 0.9
```
This spacing's a bit odd... you'd still have the multipliers aligned if you put each `return` on a new line

iamgabrielma replied:
True that, updated on https://github.com/woocommerce/woocommerce-ios/pull/14919/commits/0cddff1d6bdc2d045c957efd2cca25b7c279638e

joshheald reviewed (approved):
Looking better, thanks for the changes. Some further suggestions to improve the accessibility

I poked around with removing the limitation on the largest dynamic text size. By default, the scrollview doesn't allow you to scroll the content covered by the keyboard in to view, i.e. the button.

One way I found to resolve it was to add the KeyboardObserver back, then add the following to the scroll view.

```
                .safeAreaPadding(.bottom, keyboardObserver.keyboardHeight)
                .scrollDismissesKeyboard(.interactively)
```

`safeAreaPadding` is iOS 17 only, but with some careful use of availability checks you could fall back to the current solution (limited to `accessibility1`) for iOS 16. It might be best to do this in a view modifier... but not necessarily. 

Note that I also moved our navigation header out of the scrollview.

Here's how it looks:


https://github.com/user-attachments/assets/346b00dd-5409-4e71-af7a-755dc3cf10f1

The iOS 16 support may go away if I implement `Observation` as I intend to. There are backport packages we could use but I'd rather not, for a small percentage of devices that we intend to drop support soon.

iamgabrielma committed:
styling

iamgabrielma commented:
Thanks for the review and the comments @joshheald . The `.safeAreaPadding` recommendation looks like a great way to go, and really appreciate it, but as I mention on https://github.com/woocommerce/woocommerce-ios/pull/14919#discussion_r1928164622 I'll hold it for now just because of testing time before the freeze. I'll circle back to it as soon as possible since we may not even need to add the conditional handling depending on how the conversation about supporting iOS16 goes.

I'll go ahead and merge this one for now 👍
